src/electron/process_manager.ts

```diff
@@ -272,100 +176,130 @@ export class ConnectionManager {
+
+    this.process.launch(this.getProcessArgs());
+
+    return new Promise((resolve, reject) => {
+      // Declare success when tun2socks is running.
+      this.process.onStderr = (data?: string | Buffer) => {
+        if (data && data.toString().includes('tun2socks running')) {
+          this.process.onStderr = undefined;
+          this.notifyStop = true;
+          if (isWindows) {
+            powerMonitor.once('suspend', this.suspendListener);
+            powerMonitor.once('resume', this.resumeListener);
+          }
+          resolve();
+        }
+      };
+
+      // Reject on early exit. On routine exit, the promise will be already resolved.
+      this.process.onExit.then((code?: number) => {
+        console.log('tun2socks exited with code', code);
+        if (!this.notifyStop) {
+          return;
+        }
+        this.resolveStop();
+        reject(errors.fromErrorCode(code || errors.ErrorCode.UNEXPECTED));
+      });
+    });
   }
 
-  // Use #onExit to be notified when the process exits.
-  stop() {
-    if (!this.process) {
-      // Never started.
-      if (this.exitListener) {
-        this.exitListener();
-      }
-      return;
+  // Stops the tun2socks process. Notifies the caller when the process has exited
+  // by resolving `onceStopped` if `notify` is true.
+  stop(notify=true) {
+    if (isWindows) {
+      powerMonitor.removeListener('suspend', this.suspendListener);
+      powerMonitor.removeListener('resume', this.resumeListener);
     }
-
-    this.process.kill();
+    this.notifyStop = notify;
+    this.process.stop();
   }
 
-  set onExit(newListener: (() => void)|undefined) {
-    this.exitListener = newListener;
+  get onceStopped(): Promise<void> {
+    return this.stopped;
   }
-}
 
-class SsLocal extends ChildProcessHelper {
-  constructor(private readonly proxyPort: number) {
-    super(pathToEmbeddedBinary('shadowsocks-libev', 'ss-local'));
+  // Notifes tun2socks that network connectivity changed. Checks whehter UDP support changed,
+  // restarting if it did.
+  async networkChanged() {
+    if (this.isUdpEnabled === await this.checkConnectivity()) {
+      return;
+    }
+    this.isUdpEnabled = !this.isUdpEnabled;
+    return this.start(false);
   }
 
-  start(config: cordova.plugins.outline.ServerConfig) {
-    // ss-local -s x.x.x.x -p 65336 -k mypassword -m aes-128-cfb -l 1081 -u
-    const args = ['-l', this.proxyPort.toString()];
-    args.push('-s', config.host || '');
-    args.push('-p', '' + config.port);
-    args.push('-k', config.password || '');
-    args.push('-m', config.method || '');
-    args.push('-u');
+  private async checkConnectivity() {
+    return (new ShadowsocksConnectivity(this.config)).onceResult;
+  }
 
-    this.launch(args);
+  private suspendListener = () => {
+    console.log('system suspending');
+    // Windows: when the system suspends, tun2socks terminates due to the TAP device getting closed.
+    // Don't notify this exit.
+    this.notifyStop = false;
```

Comment:
Is there any reason to wait for tun2socks to terminate itself, rather than terminating it ourselves here?

Reply:
Previously, we weren't calling stop to avoid unregistering the suspend/resume listeners. We're now stopping the process (directly) on suspend.